File: app/services/payment_service.py
from __future__ import annotations

import os
import secrets
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Optional
from urllib.parse import urlencode

import requests
from sqlalchemy.orm import Session

from app.core.env import load_local_env
from app.models.orm import OrganizationPaymentSettings, PaymentRequest

load_local_env()


@dataclass
class CreatedPaymentLink:
    provider: str
    payment_url: str
    provider_payment_id: Optional[str] = None
    provider_session_id: Optional[str] = None
    provider_payload: Optional[dict] = None


class PaymentService:
    def __init__(self) -> None:
        self.fallback_provider = (os.getenv("ACE_PAYMENT_PROVIDER") or "mock").strip().lower()
        self.public_base_url = (os.getenv("ACE_PUBLIC_BASE_URL") or "http://localhost:8000").rstrip("/")
        self.dashboard_base_url = (os.getenv("ACE_MANAGER_DASHBOARD_URL") or "http://localhost:4400").rstrip("/")
        self.stripe_secret_key = (os.getenv("STRIPE_SECRET_KEY") or "").strip()
        self.stripe_connect_client_id = (os.getenv("STRIPE_CONNECT_CLIENT_ID") or "").strip()

    # --------------------------- payment settings ---------------------------
    def get_or_create_settings(self, db: Session, *, organization_id: int) -> OrganizationPaymentSettings:
        settings = db.query(OrganizationPaymentSettings).filter(
            OrganizationPaymentSettings.organization_id == organization_id
        ).first()
        if settings:
            return settings
        settings = OrganizationPaymentSettings(
            organization_id=organization_id,
            provider="stripe",
            mode="stripe_connect_standard",
            payments_enabled=False,
            default_currency="EUR",
            stripe_connect_status="not_connected",
        )
        db.add(settings)
        db.commit()
        db.refresh(settings)
        return settings

    def refresh_connect_status(self, db: Session, *, settings: OrganizationPaymentSettings) -> OrganizationPaymentSettings:
        if not settings.stripe_account_id or not self.stripe_secret_key:
            settings.payments_enabled = False
            settings.stripe_connect_status = "not_connected"
            settings.last_synced_at = datetime.utcnow()
            db.add(settings)
            db.commit()
            db.refresh(settings)
            return settings

        response = requests.get(
            f"https://api.stripe.com/v1/accounts/{settings.stripe_account_id}",
            auth=(self.stripe_secret_key, ""),
            timeout=20,
        )
        response.raise_for_status()
        account = response.json()
        settings.stripe_onboarding_complete = bool(account.get("details_submitted"))
        settings.stripe_details_submitted = bool(account.get("details_submitted"))
        settings.stripe_charges_enabled = bool(account.get("charges_enabled"))
        settings.stripe_payouts_enabled = bool(account.get("payouts_enabled"))
        settings.stripe_livemode = bool(account.get("livemode"))
        settings.stripe_last_error = None
        settings.last_synced_at = datetime.utcnow()
        settings.payments_enabled = bool(settings.stripe_charges_enabled)
        settings.stripe_connect_status = "connected" if settings.payments_enabled else "restricted"
        db.add(settings)
        db.commit()
        db.refresh(settings)
        return settings

    def create_connect_link(self, db: Session, *, organization_id: int) -> str:
        if not self.stripe_connect_client_id:
            raise RuntimeError("Missing STRIPE_CONNECT_CLIENT_ID")

        settings = self.get_or_create_settings(db, organization_id=organization_id)
        state = secrets.token_urlsafe(24)
        settings.stripe_oauth_state = state
        settings.stripe_connect_status = "pending"
        settings.stripe_last_error = None
        db.add(settings)
        db.commit()

        params = {
            "response_type": "code",
            "client_id": self.stripe_connect_client_id,
            "scope": "read_write",
            "state": state,
            "redirect_uri": f"{self.public_base_url}/api/public/payments/stripe/connect/callback",
            "suggested_capabilities[]": "transfers",
            "stripe_user[business_type]": "company",
        }
        return f"https://connect.stripe.com/oauth/authorize?{urlencode(params, doseq=True)}"

    def handle_connect_callback(self, db: Session, *, state: str, code: str) -> OrganizationPaymentSettings:
        settings = db.query(OrganizationPaymentSettings).filter(
            OrganizationPaymentSettings.stripe_oauth_state == state
        ).first()
        if not settings:
            raise ValueError("Invalid or expired Stripe connect state")
        if not self.stripe_secret_key:
            raise RuntimeError("Missing STRIPE_SECRET_KEY")

        response = requests.post(
            "https://connect.stripe.com/oauth/token",
            auth=(self.stripe_secret_key, ""),
            data={
                "grant_type": "authorization_code",
                "code": code,
            },
            timeout=20,
        )
        response.raise_for_status()
        data = response.json()

        settings.stripe_account_id = data.get("stripe_user_id")
        settings.stripe_access_token = data.get("access_token")
        settings.stripe_refresh_token = data.get("refresh_token")
        settings.stripe_publishable_key = data.get("stripe_publishable_key")
        settings.stripe_scope = data.get("scope")
        settings.stripe_livemode = bool(data.get("livemode"))
        settings.stripe_oauth_state = None
        settings.stripe_last_error = None
        db.add(settings)
        db.commit()
        db.refresh(settings)
        return self.refresh_connect_status(db, settings=settings)

    def mark_connect_error(self, db: Session, *, state: Optional[str], error_message: str) -> Optional[OrganizationPaymentSettings]:
        if not state:
            return None
        settings = db.query(OrganizationPaymentSettings).filter(
            OrganizationPaymentSettings.stripe_oauth_state == state
        ).first()
        if not settings:
            return None
        settings.stripe_connect_status = "error"
        settings.stripe_last_error = error_message[:1000]
        settings.stripe_oauth_state = None
        settings.payments_enabled = False
        settings.last_synced_at = datetime.utcnow()
        db.add(settings)
        db.commit()
        db.refresh(settings)
        return settings

    # --------------------------- payment requests ---------------------------
    def create_payment_request(
        self,
        db: Session,
        *,
        organization_id: int,
        sid: str,
        created_by_user_id: Optional[int],
        amount: float,
        currency: str,
        purpose: str,
        note: str = "",
        expires_in_hours: Optional[int] = 24,
    ) -> PaymentRequest:
        amount_cents = max(int(round(float(amount) * 100)), 1)
        settings = self.get_or_create_settings(db, organization_id=organization_id)
        provider = self._resolve_provider(settings)
        currency_norm = (currency or settings.default_currency or "EUR").strip().upper()

        payment_request = PaymentRequest(
            organization_id=organization_id,
            sid=sid,
            created_by_user_id=created_by_user_id,
            provider=provider,
            public_token=secrets.token_urlsafe(18),
            amount_cents=amount_cents,
            currency=currency_norm,
            purpose=(purpose or "Payment request").strip(),
            note=(note or "").strip(),
            status="draft",
            payment_url="",
            expires_at=(datetime.utcnow() + timedelta(hours=expires_in_hours)) if expires_in_hours else None,
        )
        db.add(payment_request)
        db.flush()

        link = self._build_link(payment_request, settings=settings)
        payment_request.provider = link.provider
        payment_request.payment_url = link.payment_url
        payment_request.provider_payment_id = link.provider_payment_id
        payment_request.provider_session_id = link.provider_session_id
        payment_request.provider_payload = link.provider_payload
        payment_request.status = "sent"

        db.add(payment_request)
        db.commit()
        db.refresh(payment_request)
        return payment_request

    def list_requests(self, db: Session, *, organization_id: int, sid: Optional[str] = None, limit: int = 100) -> list[PaymentRequest]:
        query = db.query(PaymentRequest).filter(PaymentRequest.organization_id == organization_id)
        if sid:
            query = query.filter(PaymentRequest.sid == sid)
        return query.order_by(PaymentRequest.created_at.desc()).limit(limit).all()

    def get_by_id(self, db: Session, *, organization_id: int, request_id: int) -> Optional[PaymentRequest]:
        return db.query(PaymentRequest).filter(
            PaymentRequest.organization_id == organization_id,
            PaymentRequest.id == request_id,
        ).first()

    def get_by_token(self, db: Session, *, public_token: str) -> Optional[PaymentRequest]:
        return db.query(PaymentRequest).filter(PaymentRequest.public_token == public_token).first()

    def mark_paid(
        self,
        db: Session,
        *,
        payment_request: PaymentRequest,
        provider_payment_id: Optional[str] = None,
        provider_session_id: Optional[str] = None,
        provider_payload: Optional[dict] = None,
    ) -> PaymentRequest:
        if payment_request.status == "paid":
            return payment_request
        payment_request.status = "paid"
        payment_request.paid_at = datetime.utcnow()
        if provider_payment_id:
            payment_request.provider_payment_id = provider_payment_id
        if provider_session_id:
            payment_request.provider_session_id = provider_session_id
        if provider_payload:
            payment_request.provider_payload = provider_payload
        db.add(payment_request)
        db.commit()
        db.refresh(payment_request)
        return payment_request

    def mark_cancelled(self, db: Session, *, payment_request: PaymentRequest) -> PaymentRequest:
        if payment_request.status == "paid":
            return payment_request
        payment_request.status = "cancelled"
        db.add(payment_request)
        db.commit()
        db.refresh(payment_request)
        return payment_request

    def verify_stripe_checkout(self, *, session_id: str, stripe_account_id: Optional[str] = None) -> Optional[dict]:
        if not self.stripe_secret_key:
            return None
        headers = {}
        if stripe_account_id:
            headers["Stripe-Account"] = stripe_account_id
        response = requests.get(
            f"https://api.stripe.com/v1/checkout/sessions/{session_id}",
            auth=(self.stripe_secret_key, ""),
            headers=headers,
            timeout=20,
        )
        if response.status_code >= 400:
            return None
        return response.json()

    # ------------------------------ internals ------------------------------
    def _resolve_provider(self, settings: OrganizationPaymentSettings) -> str:
        if settings.stripe_account_id and settings.payments_enabled and self.stripe_secret_key:
            return "stripe_connect"
        if self.stripe_secret_key:
            return "stripe_demo"
        return self.fallback_provider if self.fallback_provider in {"mock"} else "mock"

    def _build_link(self, payment_request: PaymentRequest, *, settings: OrganizationPaymentSettings) -> CreatedPaymentLink:
        if payment_request.provider == "stripe_connect" and self.stripe_secret_key and settings.stripe_account_id:
            try:
                return self._create_stripe_checkout(payment_request, settings=settings)
            except Exception as exc:
                settings.stripe_last_error = str(exc)[:1000]
                settings.stripe_connect_status = "error"
                settings.payments_enabled = False
                settings.last_synced_at = datetime.utcnow()
                raise
        if payment_request.provider == "stripe_demo" and self.stripe_secret_key:
            return self._create_platform_stripe_checkout(payment_request)
        return self._create_mock_link(payment_request)

    def _create_mock_link(self, payment_request: PaymentRequest) -> CreatedPaymentLink:
        return CreatedPaymentLink(
            provider="mock",
            payment_url=f"{self.public_base_url}/pay/{payment_request.public_token}",
            provider_payload={"mode": "mock"},
        )

    def _stripe_checkout_payload(self, payment_request: PaymentRequest) -> dict:
        success_url = (
            f"{self.public_base_url}/pay/success"
            f"?payment_request_id={payment_request.id}&session_id={{CHECKOUT_SESSION_ID}}"
        )
        cancel_url = f"{self.public_base_url}/pay/cancel?payment_request_id={payment_request.id}"
        payload = {
            "mode": "payment",
            "success_url": success_url,
            "cancel_url": cancel_url,
            "client_reference_id": str(payment_request.id),
            "metadata[payment_request_id]": str(payment_request.id),
            "metadata[sid]": payment_request.sid,
            "metadata[organization_id]": str(payment_request.organization_id),
            "line_items[0][quantity]": "1",
            "line_items[0][price_data][currency]": payment_request.currency.lower(),
            "line_items[0][price_data][unit_amount]": str(payment_request.amount_cents),
            "line_items[0][price_data][product_data][name]": payment_request.purpose,
        }
        if payment_request.note:
            payload["line_items[0][price_data][product_data][description]"] = payment_request.note[:500]
        return payload

    def _create_platform_stripe_checkout(self, payment_request: PaymentRequest) -> CreatedPaymentLink:
        response = requests.post(
            "https://api.stripe.com/v1/checkout/sessions",
            auth=(self.stripe_secret_key, ""),
            data=self._stripe_checkout_payload(payment_request),
            timeout=20,
        )
        response.raise_for_status()
        data = response.json()
        return CreatedPaymentLink(
            provider="stripe_demo",
            payment_url=data["url"],
            provider_payment_id=data.get("payment_intent"),
            provider_session_id=data.get("id"),
            provider_payload={
                "mode": "stripe_demo_checkout",
                "session_id": data.get("id"),
                "livemode": data.get("livemode", False),
            },
        )

    def _create_stripe_checkout(self, payment_request: PaymentRequest, *, settings: OrganizationPaymentSettings) -> CreatedPaymentLink:
        response = requests.post(
            "https://api.stripe.com/v1/checkout/sessions",
            auth=(self.stripe_secret_key, ""),
            headers={"Stripe-Account": settings.stripe_account_id},
            data=self._stripe_checkout_payload(payment_request),
            timeout=20,
        )
        response.raise_for_status()
        data = response.json()
        return CreatedPaymentLink(
            provider="stripe_connect",
            payment_url=data["url"],
            provider_payment_id=data.get("payment_intent"),
            provider_session_id=data.get("id"),
            provider_payload={
                "mode": "stripe_connect_checkout",
                "session_id": data.get("id"),
                "livemode": data.get("livemode", False),
                "stripe_account_id": settings.stripe_account_id,
            },
        )


service = PaymentService()

File: java-platform/src/main/java/com/ace/platform/payment/PaymentService.java
package com.ace.platform.payment;

import com.ace.platform.organization.Organization;
import com.ace.platform.organization.OrganizationRepository;
import com.ace.platform.user.User;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.IOException;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.time.Duration;
import java.time.Instant;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
@ConditionalOnProperty(name = "ace.payments.enabled", havingValue = "true", matchIfMissing = true)
public class PaymentService {

    private final OrganizationRepository organizationRepository;
    private final OrganizationPaymentSettingsRepository settingsRepository;
    private final PaymentRequestRepository paymentRequestRepository;
    private final ObjectMapper objectMapper;
    private final HttpClient httpClient;
    private final SecureRandom secureRandom = new SecureRandom();
    private final String fallbackProvider;
    private final String publicBaseUrl;
    private final String stripeSecretKey;
    private final String stripeConnectClientId;

    public PaymentService(
        OrganizationRepository organizationRepository,
        OrganizationPaymentSettingsRepository settingsRepository,
        PaymentRequestRepository paymentRequestRepository,
        ObjectMapper objectMapper,
        @Value("${ACE_PAYMENT_PROVIDER:mock}") String fallbackProvider,
        @Value("${ACE_PUBLIC_BASE_URL:http://localhost:8080}") String publicBaseUrl,
        @Value("${STRIPE_SECRET_KEY:}") String stripeSecretKey,
        @Value("${STRIPE_CONNECT_CLIENT_ID:}") String stripeConnectClientId
    ) {
        this.organizationRepository = organizationRepository;
        this.settingsRepository = settingsRepository;
        this.paymentRequestRepository = paymentRequestRepository;
        this.objectMapper = objectMapper;
        this.fallbackProvider = fallbackProvider != null ? fallbackProvider.trim().toLowerCase() : "mock";
        this.publicBaseUrl = stripTrailingSlash(publicBaseUrl != null ? publicBaseUrl.trim() : "http://localhost:8080");
        this.stripeSecretKey = stripeSecretKey != null ? stripeSecretKey.trim() : "";
        this.stripeConnectClientId = stripeConnectClientId != null ? stripeConnectClientId.trim() : "";
        this.httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build();
    }

    @Transactional
    public OrganizationPaymentSettings getOrCreateSettings(Long organizationId) {
        return settingsRepository.findByOrganizationId(organizationId)
            .orElseGet(() -> {
                Organization organization = organizationRepository.findById(organizationId)
                    .orElseThrow(() -> new IllegalArgumentException("Organization not found"));
                OrganizationPaymentSettings settings = new OrganizationPaymentSettings(organization);
                settings.setProvider("stripe");
                settings.setMode("stripe_connect_standard");
                settings.setPaymentsEnabled(false);
                settings.setDefaultCurrency("EUR");
                settings.setStripeConnectStatus("not_connected");
                return settingsRepository.save(settings);
            });
    }

    @Transactional
    public OrganizationPaymentSettings refreshConnectStatus(OrganizationPaymentSettings settings) {
        if (blank(settings.getStripeAccountId()) || blank(stripeSecretKey)) {
            settings.setPaymentsEnabled(false);
            settings.setStripeConnectStatus("not_connected");
            settings.setLastSyncedAt(Instant.now());
            return settingsRepository.save(settings);
        }

        Map<String, Object> account = stripeGet("https://api.stripe.com/v1/accounts/" + encode(settings.getStripeAccountId()), null);
        settings.setStripeOnboardingComplete(bool(account.get("details_submitted")));
        settings.setStripeDetailsSubmitted(bool(account.get("details_submitted")));
        settings.setStripeChargesEnabled(bool(account.get("charges_enabled")));
        settings.setStripePayoutsEnabled(bool(account.get("payouts_enabled")));
        settings.setStripeLivemode(bool(account.get("livemode")));
        settings.setStripeLastError(null);
        settings.setLastSyncedAt(Instant.now());
        settings.setPaymentsEnabled(settings.isStripeChargesEnabled());
        settings.setStripeConnectStatus(settings.isPaymentsEnabled() ? "connected" : "restricted");
        return settingsRepository.save(settings);
    }

    @Transactional
    public String createConnectLink(Long organizationId) {
        if (blank(stripeConnectClientId)) {
            throw new IllegalStateException("Missing STRIPE_CONNECT_CLIENT_ID");
        }
        OrganizationPaymentSettings settings = getOrCreateSettings(organizationId);
        String state = token(24);
        settings.setStripeOauthState(state);
        settings.setStripeConnectStatus("pending");
        settings.setStripeLastError(null);
        settingsRepository.save(settings);

        Map<String, String> params = new LinkedHashMap<>();
        params.put("response_type", "code");
        params.put("client_id", stripeConnectClientId);
        params.put("scope", "read_write");
        params.put("state", state);
        params.put("redirect_uri", publicBaseUrl + "/api/public/payments/stripe/connect/callback");
        params.put("suggested_capabilities[]", "transfers");
        params.put("stripe_user[business_type]", "company");
        return "https://connect.stripe.com/oauth/authorize?" + formEncode(params);
    }

    @Transactional
    public OrganizationPaymentSettings handleConnectCallback(String state, String code) {
        OrganizationPaymentSettings settings = settingsRepository.findByStripeOauthState(state)
            .orElseThrow(() -> new IllegalArgumentException("Invalid or expired Stripe connect state"));
        if (blank(stripeSecretKey)) {
            throw new IllegalStateException("Missing STRIPE_SECRET_KEY");
        }

        Map<String, String> form = new LinkedHashMap<>();
        form.put("grant_type", "authorization_code");
        form.put("code", code);
        Map<String, Object> data = stripePost("https://connect.stripe.com/oauth/token", form, null);

        settings.setStripeAccountId(string(data.get("stripe_user_id")));
        settings.setStripeAccessToken(string(data.get("access_token")));
        settings.setStripeRefreshToken(string(data.get("refresh_token")));
        settings.setStripePublishableKey(string(data.get("stripe_publishable_key")));
        settings.setStripeScope(string(data.get("scope")));
        settings.setStripeLivemode(bool(data.get("livemode")));
        settings.setStripeOauthState(null);
        settings.setStripeLastError(null);
        settingsRepository.save(settings);
        return refreshConnectStatus(settings);
    }

    @Transactional
    public Optional<OrganizationPaymentSettings> markConnectError(String state, String errorMessage) {
        if (blank(state)) return Optional.empty();
        return settingsRepository.findByStripeOauthState(state).map(settings -> {
            settings.setStripeConnectStatus("error");
            settings.setStripeLastError(truncate(errorMessage, 1000));
            settings.setStripeOauthState(null);
            settings.setPaymentsEnabled(false);
            settings.setLastSyncedAt(Instant.now());
            return settingsRepository.save(settings);
        });
    }

    @Transactional
    public PaymentRequest createPaymentRequest(
        Long organizationId,
        String sid,
        User createdByUser,
        Double amount,
        String currency,
        String purpose,
        String note,
        Integer expiresInHours
    ) {
        Organization organization = organizationRepository.findById(organizationId)
            .orElseThrow(() -> new IllegalArgumentException("Organization not found"));
        OrganizationPaymentSettings settings = getOrCreateSettings(organizationId);

        int amountCents = amountToCents(amount);
        String currencyNorm = !blank(currency) ? currency.trim().toUpperCase() : settings.getDefaultCurrency();
        String purposeNorm = !blank(purpose) ? purpose.trim() : "Payment request";
        String noteNorm = note != null ? note.trim() : "";

        PaymentRequest paymentRequest = new PaymentRequest(
            organization,
            createdByUser,
            sid,
            token(18),
            amountCents,
            currencyNorm,
            purposeNorm,
            noteNorm
        );
        paymentRequest.setProvider(resolveProvider(settings));
        paymentRequest.setStatus("draft");
        if (expiresInHours != null && expiresInHours > 0) {
            paymentRequest.setExpiresAt(Instant.now().plus(Duration.ofHours(expiresInHours)));
        }
        paymentRequest = paymentRequestRepository.save(paymentRequest);

        CreatedPaymentLink link = buildLink(paymentRequest, settings);
        paymentRequest.setProvider(link.provider());
        paymentRequest.setPaymentUrl(link.paymentUrl());
        paymentRequest.setProviderPaymentId(link.providerPaymentId());
        paymentRequest.setProviderSessionId(link.providerSessionId());
        paymentRequest.setProviderPayload(link.providerPayload());
        paymentRequest.setStatus("sent");
        return paymentRequestRepository.save(paymentRequest);
    }

    @Transactional(readOnly = true)
    public List<PaymentRequest> listRequests(Long organizationId, String sid, int limit) {
        PageRequest page = PageRequest.of(0, Math.max(1, Math.min(limit, 500)));
        if (blank(sid)) {
            return paymentRequestRepository.findByOrganizationIdOrderByCreatedAtDesc(organizationId, page);
        }
        return paymentRequestRepository.findByOrganizationIdAndSidOrderByCreatedAtDesc(organizationId, sid, page);
    }

    @Transactional(readOnly = true)
    public Optional<PaymentRequest> getById(Long organizationId, Long requestId) {
        return paymentRequestRepository.findByOrganizationIdAndId(organizationId, requestId);
    }

    @Transactional(readOnly = true)
    public Optional<PaymentRequest> getRequestById(Long requestId) {
        return paymentRequestRepository.findById(requestId);
    }

    @Transactional(readOnly = true)
    public Optional<PaymentRequest> getByPublicToken(String publicToken) {
        return paymentRequestRepository.findByPublicToken(publicToken);
    }

    @Transactional(readOnly = true)
    public Optional<OrganizationPaymentSettings> findSettingsByStripeAccountId(String stripeAccountId) {
        return settingsRepository.findByStripeAccountId(stripeAccountId);
    }

    @Transactional
    public PaymentRequest markPaid(PaymentRequest paymentRequest, String providerPaymentId, String providerSessionId, Map<String, Object> providerPayload) {
        if ("paid".equals(paymentRequest.getStatus())) {
            return paymentRequest;
        }
        paymentRequest.setStatus("paid");
        paymentRequest.setPaidAt(Instant.now());
        if (!blank(providerPaymentId)) paymentRequest.setProviderPaymentId(providerPaymentId);
        if (!blank(providerSessionId)) paymentRequest.setProviderSessionId(providerSessionId);
        if (providerPayload != null && !providerPayload.isEmpty()) paymentRequest.setProviderPayload(providerPayload);
        return paymentRequestRepository.save(paymentRequest);
    }

    @Transactional
    public PaymentRequest markCancelled(PaymentRequest paymentRequest) {
        if ("paid".equals(paymentRequest.getStatus())) {
            return paymentRequest;
        }
        paymentRequest.setStatus("cancelled");
        return paymentRequestRepository.save(paymentRequest);
    }

    @Transactional(readOnly = true)
    public Map<String, Object> verifyStripeCheckout(String sessionId, String stripeAccountId) {
        if (blank(stripeSecretKey) || blank(sessionId)) {
            return null;
        }
        return stripeGet("https://api.stripe.com/v1/checkout/sessions/" + encode(sessionId), blank(stripeAccountId) ? null : Map.of("Stripe-Account", stripeAccountId));
    }

    public Map<String, Object> toSettingsPayload(OrganizationPaymentSettings settings) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("id", settings.getId());
        out.put("organization_id", settings.getOrganization().getId());
        out.put("provider", settings.getProvider());
        out.put("mode", settings.getMode());
        out.put("payments_enabled", settings.isPaymentsEnabled());
        out.put("default_currency", settings.getDefaultCurrency());
        out.put("stripe_account_id", settings.getStripeAccountId());
        out.put("stripe_connect_status", settings.getStripeConnectStatus());
        out.put("stripe_onboarding_complete", settings.isStripeOnboardingComplete());
        out.put("stripe_details_submitted", settings.isStripeDetailsSubmitted());
        out.put("stripe_charges_enabled", settings.isStripeChargesEnabled());
        out.put("stripe_payouts_enabled", settings.isStripePayoutsEnabled());
        out.put("stripe_publishable_key", settings.getStripePublishableKey());
        out.put("stripe_scope", settings.getStripeScope());
        out.put("stripe_livemode", settings.isStripeLivemode());
        out.put("stripe_last_error", settings.getStripeLastError());
        out.put("last_synced_at", settings.getLastSyncedAt() != null ? settings.getLastSyncedAt().toString() : null);
        out.put("created_at", settings.getCreatedAt() != null ? settings.getCreatedAt().toString() : null);
        out.put("updated_at", settings.getUpdatedAt() != null ? settings.getUpdatedAt().toString() : null);
        return out;
    }

    public Map<String, Object> toPaymentRequestPayload(PaymentRequest item) {
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("id", item.getId());
        out.put("organization_id", item.getOrganization().getId());
        out.put("sid", item.getSid());
        out.put("created_by_user_id", item.getCreatedByUser() != null ? item.getCreatedByUser().getId() : null);
        out.put("provider", item.getProvider());
        out.put("provider_payment_id", item.getProviderPaymentId());
        out.put("provider_session_id", item.getProviderSessionId());
        out.put("public_token", item.getPublicToken());
        out.put("amount_cents", item.getAmountCents());
        out.put("currency", item.getCurrency());
        out.put("purpose", item.getPurpose());
        out.put("note", item.getNote());
        out.put("status", item.getStatus());
        out.put("payment_url", item.getPaymentUrl());
        out.put("expires_at", item.getExpiresAt() != null ? item.getExpiresAt().toString() : null);
        out.put("paid_at", item.getPaidAt() != null ? item.getPaidAt().toString() : null);
        out.put("provider_payload", item.getProviderPayload() != null ? item.getProviderPayload() : Map.of());
        out.put("created_at", item.getCreatedAt() != null ? item.getCreatedAt().toString() : null);
        out.put("updated_at", item.getUpdatedAt() != null ? item.getUpdatedAt().toString() : null);
        return out;
    }

    public String dashboardUrlForOrganization(String orgSlug) {
        return "/" + orgSlug + "/dashboard?tab=payments";
    }

    private String resolveProvider(OrganizationPaymentSettings settings) {
        if (!blank(settings.getStripeAccountId()) && settings.isPaymentsEnabled() && !blank(stripeSecretKey)) {
            return "stripe_connect";
        }
        if (!blank(stripeSecretKey)) {
            return "stripe_demo";
        }
        return "mock".equals(fallbackProvider) ? "mock" : "mock";
    }

    private CreatedPaymentLink buildLink(PaymentRequest paymentRequest, OrganizationPaymentSettings settings) {
        if ("stripe_connect".equals(paymentRequest.getProvider()) && !blank(stripeSecretKey) && !blank(settings.getStripeAccountId())) {
            try {
                return createStripeCheckout(paymentRequest, settings.getStripeAccountId());
            } catch (RuntimeException ex) {
                settings.setStripeLastError(truncate(ex.getMessage(), 1000));
                settings.setStripeConnectStatus("error");
                settings.setPaymentsEnabled(false);
                settings.setLastSyncedAt(Instant.now());
                settingsRepository.save(settings);
                throw ex;
            }
        }
        if ("stripe_demo".equals(paymentRequest.getProvider()) && !blank(stripeSecretKey)) {
            return createStripeCheckout(paymentRequest, null);
        }
        return new CreatedPaymentLink("mock", publicBaseUrl + "/pay/" + paymentRequest.getPublicToken(), null, null, Map.of("mode", "mock"));
    }

    private CreatedPaymentLink createStripeCheckout(PaymentRequest paymentRequest, String stripeAccountId) {
        Map<String, String> form = stripeCheckoutPayload(paymentRequest);
        Map<String, String> headers = stripeAccountId != null ? Map.of("Stripe-Account", stripeAccountId) : null;
        Map<String, Object> data = stripePost("https://api.stripe.com/v1/checkout/sessions", form, headers);
        return new CreatedPaymentLink(
            stripeAccountId != null ? "stripe_connect" : "stripe_demo",
            string(data.get("url")),
            string(data.get("payment_intent")),
            string(data.get("id")),
            stripeAccountId != null
                ? Map.of("mode", "stripe_connect_checkout", "session_id", string(data.get("id")), "livemode", bool(data.get("livemode")), "stripe_account_id", stripeAccountId)
                : Map.of("mode", "stripe_demo_checkout", "session_id", string(data.get("id")), "livemode", bool(data.get("livemode")))
        );
    }

    private Map<String, String> stripeCheckoutPayload(PaymentRequest paymentRequest) {
        Map<String, String> payload = new LinkedHashMap<>();
        payload.put("mode", "payment");
        payload.put("success_url", publicBaseUrl + "/pay/success?payment_request_id=" + paymentRequest.getId() + "&session_id={CHECKOUT_SESSION_ID}");
        payload.put("cancel_url", publicBaseUrl + "/pay/cancel?payment_request_id=" + paymentRequest.getId());
        payload.put("client_reference_id", String.valueOf(paymentRequest.getId()));
        payload.put("metadata[payment_request_id]", String.valueOf(paymentRequest.getId()));
        payload.put("metadata[sid]", paymentRequest.getSid());
        payload.put("metadata[organization_id]", String.valueOf(paymentRequest.getOrganization().getId()));
        payload.put("line_items[0][quantity]", "1");
        payload.put("line_items[0][price_data][currency]", paymentRequest.getCurrency().toLowerCase());
        payload.put("line_items[0][price_data][unit_amount]", String.valueOf(paymentRequest.getAmountCents()));
        payload.put("line_items[0][price_data][product_data][name]", paymentRequest.getPurpose());
        if (!blank(paymentRequest.getNote())) {
            payload.put("line_items[0][price_data][product_data][description]", truncate(paymentRequest.getNote(), 500));
        }
        return payload;
    }

    private Map<String, Object> stripeGet(String url, Map<String, String> extraHeaders) {
        try {
            HttpRequest.Builder builder = HttpRequest.newBuilder().uri(URI.create(url)).timeout(Duration.ofSeconds(20)).GET();
            builder.header("Authorization", "Basic " + basicAuthValue());
            if (extraHeaders != null) {
                extraHeaders.forEach(builder::header);
            }
            HttpResponse<String> response = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() >= 400) {
                throw new IllegalStateException("Stripe returned status " + response.statusCode() + ": " + response.body());
            }
            return objectMapper.readValue(response.body(), new TypeReference<>() {});
        } catch (IOException | InterruptedException ex) {
            if (ex instanceof InterruptedException) Thread.currentThread().interrupt();
            throw new IllegalStateException("Stripe request failed", ex);
        }
    }

    private Map<String, Object> stripePost(String url, Map<String, String> form, Map<String, String> extraHeaders) {
        try {
            HttpRequest.Builder builder = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofSeconds(20))
                .header("Authorization", "Basic " + basicAuthValue())
                .header("Content-Type", "application/x-www-form-urlencoded")
                .POST(HttpRequest.BodyPublishers.ofString(formEncode(form)));
            if (extraHeaders != null) {
                extraHeaders.forEach(builder::header);
            }
            HttpResponse<String> response = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() >= 400) {
                throw new IllegalStateException("Stripe returned status " + response.statusCode() + ": " + response.body());
            }
            return objectMapper.readValue(response.body(), new TypeReference<>() {});
        } catch (IOException | InterruptedException ex) {
            if (ex instanceof InterruptedException) Thread.currentThread().interrupt();
            throw new IllegalStateException("Stripe request failed", ex);
        }
    }

    private int amountToCents(Double amount) {
        BigDecimal value = BigDecimal.valueOf(amount != null ? amount : 0.0d).setScale(2, RoundingMode.HALF_UP);
        int cents = value.multiply(BigDecimal.valueOf(100)).intValue();
        return Math.max(cents, 1);
    }

    private String token(int bytes) {
        byte[] raw = new byte[bytes];
        secureRandom.nextBytes(raw);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(raw);
    }

    private String formEncode(Map<String, String> form) {
        StringBuilder out = new StringBuilder();
        boolean first = true;
        for (Map.Entry<String, String> entry : form.entrySet()) {
            if (!first) out.append('&');
            first = false;
            out.append(encode(entry.getKey())).append('=').append(encode(entry.getValue()));
        }
        return out.toString();
    }

    private String basicAuthValue() {
        return Base64.getEncoder().encodeToString((stripeSecretKey + ":").getBytes(StandardCharsets.UTF_8));
    }

    private String encode(String value) {
        return URLEncoder.encode(value != null ? value : "", StandardCharsets.UTF_8);
    }

    private String string(Object value) {
        return value != null ? String.valueOf(value) : null;
    }

    private boolean bool(Object value) {
        if (value instanceof Boolean b) return b;
        if (value instanceof Number n) return n.intValue() != 0;
        if (value instanceof String s) return "true".equalsIgnoreCase(s) || "1".equals(s);
        return false;
    }

    private boolean blank(String value) {
        return value == null || value.isBlank();
    }

    private String stripTrailingSlash(String value) {
        return value.endsWith("/") ? value.substring(0, value.length() - 1) : value;
    }

    private String truncate(String value, int max) {
        if (value == null) return null;
        return value.length() <= max ? value : value.substring(0, max);
    }

    public record CreatedPaymentLink(
        String provider,
        String paymentUrl,
        String providerPaymentId,
        String providerSessionId,
        Map<String, Object> providerPayload
    ) {
    }
}
